main.go

```diff
@@ -49,75 +39,70 @@ func failf(format string, v ...interface{}) {
-	if err != nil {
-		return false, formatErr(sys, err)
-	}
+	go func() {
+		time.AfterFunc(timeout, func() {
+			doneChan <- WaitForBootCompleteResult{Error: errors.New("timeout")}
+		})
+	}()
 
-	init, err := cmdRunner.RunCommandWithTimeout(adbShellCommand(androidHome, serial, "getprop init.svc.bootanim"))
-	if err != nil {
-		return false, formatErr(init, err)
-	}
+	go func() {
+		out, err := adbManager.WaitForDeviceThenShellCmd(serial, nil, "getprop sys.boot_completed").RunAndReturnTrimmedCombinedOutput()
+		if err != nil {
+			fmt.Println(out)
+			doneChan <- WaitForBootCompleteResult{Error: err}
+			return
+		}
 
-	return dev == "1" && sys == "1" && init == "stopped", nil
-}
+		lines := strings.Split(out, "\n")
+		for _, line := range lines {
+			if strings.TrimSpace(line) == "1" {
+				doneChan <- WaitForBootCompleteResult{Booted: true}
+				return
+			}
+		}
+
+		doneChan <- WaitForBootCompleteResult{Booted: false}
+	}()
 
-func terminateADBServer(androidHome string) error {
-	name, args := adbCommand(androidHome, "", "kill-server")
-	_, err := cmdRunner.RunCommandWithTimeout(name, args)
-	return err
+	return doneChan
 }
 
-func checkEmulatorBootState(androidHome, emulatorSerial string, timeout time.Duration) error {
+func waitForDevice(adb *adbmanager.Model, emulatorSerial string, timeout time.Duration) error {
 	startTime := clock.Now()
 
 	for {
-		logger.Printf("> Checking if boot completed...")
-
-		booted, err := isDeviceBooted(androidHome, emulatorSerial)
-		if err != nil {
-			switch {
-			case strings.Contains(err.Error(), "daemon not running; starting now at"):
-				logger.Warnf("adb daemon being restarted")
-				logger.Printf(err.Error())
-			case err == errTimedOut:
-				logger.Warnf("Running command timed out, retry...")
-				if err := terminateADBServer(androidHome); err != nil {
-					if err != errTimedOut {
-						return fmt.Errorf("unable to kill ADB daemon, error: %s", err)
-					}
-					logger.Warnf("killing ADB daemon timed out")
-				}
-			case err != nil:
-				return fmt.Errorf("failed to check emulator boot status: %s", err)
+		logger.Printf("Waiting for emulator boot...")
+
+		bootCompleteChan := getBootCompleteEvent(adb, emulatorSerial, timeout)
+		result := <-bootCompleteChan
+		switch {
+		case result.Error != nil:
+			logger.Warnf("Failed to check emulator boot status: %s", result.Error)
+			logger.Warnf("Killing ADB server before retry...")
+			if err := adb.KillServerCmd(nil).Run(); err != nil {
+				return fmt.Errorf("failed to terminate adb server: %s", err)
 			}
+		case result.Booted:
+			logger.Donef("Device boot completed in %d seconds", time.Since(startTime)/time.Second)
+			return nil
 		}
 
-		if booted {
-			break
-		}
-
-		if clock.Since(startTime) >= timeout {
-			return fmt.Errorf("waiting for emulator boot timed out after %d seconds", timeout)
+		if time.Now().After(startTime.Add(timeout)) {
```

Comment:
NOTE: this way the function might time out later than `startTime.Add(timeout)`, because this check is only happening after we got a result from `getBootCompleteEvent`.

To make the timeout more accurate, we could put this check into a separate go routine at the beginning of the function.

Reply:
Also, worst case we only do 1 turn of the boot complete check, as `getBootCompleteEvent` worst-case running for the passed overall `timeout`.

What about giving less time (by default 5min) to the `getBootCompleteEvent`?

Reply:
I would be very careful with introducing any extra complexity to this async logic, that's why I initially didn't add another goroutine that enforces the timeout in the most accurate way. I think it's acceptable that the step times out a few seconds after the defined timeout in rare cases (the worst case timeout is `getBootCompleteEvent()` + `adb kill-server`, right?), but of course I can change my mind. I just want to avoid complexity in this already complex code because it will be a nightmare to debug if we introduce a bug now or later.